pkg/collector/corechecks/windows_event_log/check.go

```diff
@@ -0,0 +1,294 @@
+type Check struct {
+	// check
+	core.CheckBase
+	config *Config
+
+	fetchEventsLoopWaiter sync.WaitGroup
+	fetchEventsLoopStop   chan struct{}
+
+	includedMessages []*regexp.Regexp
+	excludedMessages []*regexp.Regexp
+
+	// event metrics
+	eventPriority agentEvent.EventPriority
+
+	// event log
+	session             evtsession.Session
+	sub                 evtsubscribe.PullSubscription
+	evtapi              evtapi.API
+	systemRenderContext evtapi.EventRenderContextHandle
+	userRenderContext   evtapi.EventRenderContextHandle
+	bookmarkSaver       *bookmarkSaver
+}
+
+// Run updates sender stats, restarts the subscription if it failed, and saves the bookmark.
+// The main event collection logic runs continuously in the background, not during Run().
+func (c *Check) Run() error {
+	sender, err := c.GetSender()
+	if err != nil {
+		return err
+	}
+	// Necessary for check stats to be calculated (number of events collected, etc)
+	// Since events are collected in the background, this will update stats with the
+	// count of events collected since the last Run() call.
+	defer sender.Commit()
+
+	// Start/Restart the subscription if it is not running
+	if !c.sub.Running() {
+		// starts the event collection in the background.
+		err := c.startSubscription()
+		if err != nil {
+			return err
+		}
+	}
+
+	// save/persist bookmark on each check run
+	// check.Cancel() may not be called (like in agent check command), which means
+	// if the events read are less than bookmark_frequency then a bookmark won't
+	// be saved before the process exits. saving here, too, gives us a good time periodic
+	// save/persist in addition to the count periodic bookmark_frequency option.
+	err = c.bookmarkSaver.saveLastBookmark()
+	if err != nil {
+		c.Warnf(err.Error())
+	}
+
+	return nil
+}
+
+func (c *Check) fetchEventsLoop(sender sender.Sender) {
+	defer c.fetchEventsLoopWaiter.Done()
+
+	// Always stop the subscription when the loop ends.
+	// The check will start the subscription and this loop again next time it runs.
+	defer c.sub.Stop()
+
+	// Save the bookmark at the end of the loop, regardless of the bookmarkFrequency
+	defer func() {
+		err := c.bookmarkSaver.saveLastBookmark()
+		if err != nil {
+			c.Warnf(err.Error())
+		}
+	}()
+
+	// Fetch new events
+	for {
+		select {
+		case <-c.fetchEventsLoopStop:
+			log.Debug("stopping subscription")
+			return
+		case events, ok := <-c.sub.GetEvents():
+			if !ok {
+				// The channel is closed, this indicates an error or that sub.Stop() was called
+				// Use sub.Error() to get the error, if any.
+				err := c.sub.Error()
+				if err != nil {
+					log.Errorf("event subscription stopped with error: %v", err)
+				} else {
+					log.Debug("event subscription stopped")
+				}
+				return
+			}
+			for _, event := range events {
+				// Submit Datadog Event
+				c.submitEvent(sender, event)
+
+				// bookmarkSaver manages whether or not to save/persist the bookmark
+				err := c.bookmarkSaver.updateBookmark(event)
```

Comment:
Would it make sense to call once at the end, once we processed the batch, to avoid issuing multiple writes to disk in a short succession?

Reply:
the `bookmark_frequency` option allows the customer to control how often the bookmark is written to disk